sql-plugin/src/main/scala/com/nvidia/spark/rapids/ShimLoader.scala

```diff
@@ -16,53 +16,168 @@
 
 package com.nvidia.spark.rapids
 
-import java.util.ServiceLoader
+import java.net.URL
 
-import scala.collection.JavaConverters._
-
-import org.apache.spark.{SPARK_BUILD_USER, SPARK_VERSION}
+import org.apache.spark.{SPARK_BUILD_USER, SPARK_VERSION, SparkConf}
 import org.apache.spark.internal.Logging
+import org.apache.spark.sql.rapids.VisibleShuffleManager
+import org.apache.spark.util.{MutableURLClassLoader, ParentClassLoader}
+
+/*
+    Plugin jar uses non-standard class file layout. It consists of three types of areas
+    1. a few publicly documented classes in the conventional layout at the top
+    2. a large fraction of classes whose bytecode is identical under all supported Spark versions
+       in spark3xx-common
+    3. a smaller fraction of classes that differ under one of the supported Spark versions
+
+    com/nvidia/spark/SQLPlugin.class
+
+    spark3xx-common/com/nvidia/spark/rapids/CastExprMeta.class
+
+    spark301/org/apache/spark/sql/rapids/GpuUnaryMinus.class
+    spark311/org/apache/spark/sql/rapids/GpuUnaryMinus.class
+    spark320/org/apache/spark/sql/rapids/GpuUnaryMinus.class
+
+    Each shim can see a consistent parallel world without conflicts by referencing
+    only one conflicting directory.
+
+    E.g., Spark 3.2.0 Shim will use
+
+    jar:file:/home/spark/rapids-4-spark_2.12-21.10.jar!/spark3xx-common/
+    jar:file:/home/spark/rapids-4-spark_2.12-21.10.jar!/spark320/
+
+    Spark 3.1.1 will use
+
+    jar:file:/home/spark/rapids-4-spark_2.12-21.10.jar!/spark3xx-common/
+    jar:file:/home/spark/rapids-4-spark_2.12-21.10.jar!/spark311/
 
+    Using these Jar URL's allows referencing different bytecode produced from identical sources
+    by incompatible Scala / Spark dependencies.
+ */
 object ShimLoader extends Logging {
-  private var shimProviderClass: String = null
-  private var sparkShims: SparkShims = null
+  logDebug(s"ShimLoader object instance: ${this} loaded by ${getClass.getClassLoader}")
+  private val shimRootURL = {
+    val thisClassFile = getClass.getName.replace(".", "/") + ".class"
+    val url = getClass.getClassLoader.getResource(thisClassFile)
+    val urlStr = url.toString
+    val rootUrlStr = urlStr.substring(0, urlStr.length - thisClassFile.length)
+    new URL(rootUrlStr)
+  }
+
+  private val shimCommonURL = new URL(s"${shimRootURL.toString}spark3xx-common/")
+
+  @volatile private var shimProviderClass: String = _
+  @volatile private var sparkShims: SparkShims = _
+  @volatile private var shimURL: URL = _
+  @volatile private var pluginClassLoader: ClassLoader = _
+
+  // REPL-only logic
+  @volatile private var tmpClassLoader: MutableURLClassLoader = _
+
+  def shimId: String = shimIdFromPackageName(shimProviderClass)
+
+  def getRapidsShuffleManagerClass: String = {
+    findShimProvider()
```

Comment:
same here comment why called since not getting output and not using anything from It that is obvious